Before you turn this problem in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel$\rightarrow$Restart) and then **run all cells** (in the menubar, select Cell$\rightarrow$Run All).

Make sure you fill in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE", as well as your name and collaborators below:

In [139]:
NAME = "Michael Cornelisse"
STUDENT_NUMBER = "s1059020"
COLLABORATOR_NAME = "Nienke Helmers"
COLLABORATOR_STUDENT_NUMBER = "s1016904"

---

# Assignment 5: Association mining

## Objective of this assignment
The overall objective is to understand how frequent itemsets can be extracted by
the Apriori algorithm and be able to calculate and interpret association rules in terms of support and confidence.

## ** Important: ** When handing in your homework:
+ Hand in the notebook (and nothing else) named as follows: StudentName1_snumber_StudentName2_snumber.ipynb
+ Provide clear and complete answers to the questions below under a separate header (not hidden somewhere in your source code), and make sure to explain your answers / motivate your choices. Add Markdown cells where necessary.
+ Source code, output graphs, derivations, etc., should be included in the notebook.
+ Hand-in: upload to Blackboard.
+ Include name, student number, assignment (especially in filenames)!
+ When working in pairs only one of you should upload the assignment, and report the name of your partner in your filename.
+ For problems or questions: use the BB discussion board or email the student assistants.


## Advised Reading and Exercise Material
**The following reading material is recommended:**

- Pang-Ning Tan, Michael Steinbach, and Vipin Kumar, *Introduction to Data Mining*, section 6.


## Additional Tools
For this exercise you will need to load the provided *apriorimining.py* script. 


##  5.1 Association mining for course data 
We will use the Apriori algorithm to automatically mine for associations. The Apriori algorithm is an adapted version of the script found here: https://github.com/nalinaksh/Association-Rule-Mining-Python

Check out the script and doc and check if you understand how the association rules are computed. 




#### 5.1.1

(0.25 points) Look at the data file `data/courses.txt` by **using Python**. The data is represented in Table 1. Inspect the file `data/courses.txt` and make sure you understand how the data in Table 1 is stored in the text file. Explain how the table is a representation of the text file. What information that we would need for complete reconstruction of the table is missing from the text file?

##### Table 1
|#  |   History |Math| Biology| Spanish | Economics| Physics | Chemistry | English  |  
| :-------------: |:-------------:| :-----------:| :----------:| :------------:|:-------------:| :------------:|  :-------------: | :-------------: |
|student 1 | 0| 1 | 0 | 0 | 1| 1 |1 |1   
|student 2 | 1| 1 | 1 | 0 | 0| 1 |1 |1   
|student 3 | 0| 1 | 0 | 1 | 0| 1 |0 |1   
|student 4 | 0| 0 | 1 | 0 | 0| 1 |1 |0   
|student 5 | 0| 1 | 0 | 0 | 0| 1 |1 |0        
|student 6 | 0| 1 | 1 | 0 | 0| 1 |1 |1   


In [147]:
##Answer to question 5.1.1
import pandas as pd
data = open('data/courses.txt')
print(data.read())

2,5,6,7,8
1,2,3,6,7,8
2,4,6,8
3,6,7
2,6,7
2,3,6,7,8



Every row of the text file courses represents a student. Each row contains the index numbers of the courses that student follows starting from 1: history, math, biology, Spanish, economics, physics, chemistry and English. In the table, when a student follows the course, this is indicated by a 1, if not, there is a 0.

To reconstruct the table from the text file, the row of students and the row of courses is needed.

#### 5.1.2
(0.75 points) We will analyze the data in Table 1  using the function `associationmining.generate_association_rules()` from the toolbox. Analyze the data with $ minsupport  \geq 80 \% $ and $ minconfidence \geq 100 \%$.What
are the generated association rules? What kind of conclusions can you make based on these association rules about the subjects that students took? You can optionally provide `generate_association_rules` a dictionary `names` translating the integers in the file to the names in the table.  


In [141]:
##Answer to question 5.1.2
import toolbox.apriorimining as a

file_name = 'data/courses.txt'
courses_support = 80
courses_confidence = 100

courses = {'1': 'History', 
           '2': 'Math', 
           '3':'Biology', 
           '4':'Spanish', 
           '5':'Economics', 
           '6': 'Physics', 
           '7': 'Chemistry', 
           '8': 'English'}  

a.generate_association_rules(file_name, courses_support, courses_confidence, names=courses)

---------------TOP 10 FREQUENT 1-ITEMSET-------------------------
set= { Physics },  sup= 100.0
set= { Math },  sup= 83.33
set= { Chemistry },  sup= 83.33
-----------------------------------------------------------------
-------TOP 10 (or less) FREQUENT 2-ITEMSET------------------------
set= { Math, Physics },  sup= 83.33
set= { Chemistry, Physics },  sup= 83.33
------------------------------------------------------------------
---------------------ASSOCIATION RULES------------------
--------------------------------------------------------
Rule #1: {  } ==> { Physics }, sup= 100.00, conf= 100.00

Rule #2: { Math } ==> { Physics }, sup= 83.33, conf= 100.00

Rule #3: { Chemistry } ==> { Physics }, sup= 83.33, conf= 100.00
--------------------------------------------------------


The generated rules indicate that: everyone takes the physics course, everyone who has math also takes physics those who take chemistry also take physics The conclusion from these rules is that everyone takes physics, and most of them also take either chemistry or math.

##  5.2 Association mining for MovieLens data 
  
  
In this part of the exercise we consider a Market Basket data set containing 943 users purchases of 1682 movies. A total of 100,000 movies
have been purchased.The data set is called MovieLens100K and is provided by http://www.grouplens.org/node/73, see also the readme `MovieLensData.txt` in the data folder. The data currently considered is not the original data but modified for the apriori algorithm.

#### 5.2.1
(0.25 points) The MovieLens data is stored in the files `data/MovieLensData.txt` and `data/u.item`. Inspect the files **using Python** to see how the data is stored. How do these files relate? 
  

In [142]:
##Answer to question 5.2.1
movie_file = open('data/MovieLensData.txt')
item_file = open('data/u.item')

# Only read the first three lines of the file
print('MovieLensData:')
for i in range(3):
    print(movie_file.readline())
print('u')
for i in range(3):
    print(item_file.readline())

MovieLensData:
1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131,132,133,134,135,136,137,138,139,140,141,142,143,144,145,146,147,148,149,150,151,152,153,154,155,156,157,158,159,160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187,188,189,190,191,192,193,194,195,196,197,198,199,200,201,202,203,204,205,206,207,208,209,210,211,212,213,214,215,216,217,218,219,220,221,222,223,224,225,226,227,228,229,230,231,232,233,234,235,236,237,238,239,240,241,242,243,244,245,246,247,248,249,250,251,252,253,254,255,256,257,258,259,260,261,262,263,264,265,266,267,268,269,270,271,272

1,10

The data inside MovieLensData.txt appears to be related to the data stored in u.item.

The MovieLensData indicates which movies were in each purchase, while the u.item data stores the names of the movies and their properties (id, name, year of publishing, identifier and genres).

#### 5.2.2 
(0.75 point) Find association rules using the function below with $ minsupport  \geq 30 \% $ and $ minconfidence \geq 80 \%$. You can use the `max_rules` argument to set a maximum number of potential rules to print. Use the `movie_names` dict as the optional `names` argument in `generate_association_rules` to interpret the results. What are the associations with strongest confidence? Do these associations make sense? Explain.

In [143]:
from toolbox.load_movie_names import load_movie_names
movie_names = load_movie_names()

##Answer to question 5.2.2
file_name = 'data/MovieLensData.txt'
movie_support = 30
movie_confidence = 80

a.generate_association_rules(file_name, movie_support, movie_confidence, names=movie_names)

---------------TOP 10 FREQUENT 1-ITEMSET-------------------------
set= { Star Wars (1977) },  sup= 61.82
set= { Contact (1997) },  sup= 53.98
set= { Fargo (1996) },  sup= 53.87
set= { Return of the Jedi (1983) },  sup= 53.76
set= { Liar Liar (1997) },  sup= 51.43
set= { English Patient, The (1996) },  sup= 51.01
set= { Scream (1996) },  sup= 50.69
set= { Toy Story (1995) },  sup= 47.93
set= { Air Force One (1997) },  sup= 45.71
set= { Independence Day (ID4) (1996) },  sup= 45.49
-----------------------------------------------------------------
-------TOP 10 (or less) FREQUENT 2-ITEMSET------------------------
set= { Return of the Jedi (1983), Star Wars (1977) },  sup= 50.9
set= { Fargo (1996), Star Wars (1977) },  sup= 41.78
set= { Star Wars (1977), Toy Story (1995) },  sup= 40.4
set= { Raiders of the Lost Ark (1981), Star Wars (1977) },  sup= 40.3
set= { Independence Day (ID4) (1996), Star Wars (1977) },  sup= 38.39
set= { Godfather, The (1972), Star Wars (1977) },  sup= 37.86
set= { 

The association rules with the mosst confidence are: Rule #73: { The Empire Strikes Back, Raiders of the Lost Ark, Return of the Jedi } ==> { Star Wars } Rule #50: { The Empire Strikes Back, Return of the Jedi } ==> { Star Wars } Rule #32: { Pulp Fiction, Return of the Jedi } ==> { Star Wars } Generally this makes sense, because most of the movies are related to each other as prequel-sequel : Star Wars, The empire strikes back and Return of the Jedi. The other two movies (Raiders of the Lost Ark and Pulp Fiction) are also very well know, which would mean they have been purchased more frequently and thus contribute also to the association rules.

#### 5.2.3 
(1 point) Which movies have been watched by the most users? There are only few rules with more than three items. Why?

Star Wrs, Contact and Fargo have been watched the most, indicated by their highest support.

There are only a few rules with more than three items because of the many movie entries there are and their popularity. Many of the movies have probably been purchased only a few times, not enough to generate a conformity rule. There are only a few movies that have been purchased enough times so their rules to exceed the support and confidence threshold.

#### 5.2.4
(0.5 points) Often we are interested in rules with high confidence. Is it possible for
itemsets to have very low support but still have a very high confidence?

Suport gives the fraction of transactions that contain all items in two itemset, while the confidence indicates how often items of one itemset appear in itemsets that contain another itemset.
This means that the support is always lower than the confidence measure, because the chances of all items of the two itemsets appearing in the same set is lower than a part of one itemset and the whole of another. However, if the support is extremely low, the confidence cannot be very high, as this would indicate that there are no sets containing all items, automatically lowering the confidence measure.

## 5.3 Calculating support, confidence and interest

Calculate these measures and write down how you computed things, not just the answers. You can use Latex syntax for writing your answers.


#### 5.3.1
 Suppose we have market basket data consisting of 100 transactions and 20 items. The support for item $ \text{a} = 45 \%$, the support for item $ \text{b} = 80 \%$ and the support for itemset $ \text{ {a,b }} = 30 \%$. Let the support and confidence thresholds be 20$ \%$ and 60$ \%$, respectively.
  
1. (0.5 points) Compute the confidence of the association rule $ \text{ {a } } \rightarrow   \text{{b }} $. Is the rule interesting according to the confidence measure?

2. (0.5 points) Compute the interest measure (or lift, see slide 44 of chapter 6) for the association pattern $ \text{ {a,b}}$. Describe the nature of the relationship between item $ \text{a}$ and item $ \text{b}$  in terms of the itemset measure.
3. (1 points) What conclusion can you draw from the results of parts (1) and (2)?

#### 1
$$
\text{confidence } = \frac{ \{a, b\}}{ \{a\}} = \frac{30}{45} = 0.667\\
\text{confidence } a \rightarrow b = 67 \%\, \text{Which is above the threshold, so its an interesting measure.}
$$

#### 2
$$
\text{lift } = \frac{P(Y|X)}{P(Y)}, P(Y|X)= \frac{P(X, Y)}{P(X)}\\
P(X|Y) = \frac{ \{a, b\}}{ \{a\}} = 0.667\\
\text{lift } = \frac{0.667}{0.80} = 0.833\\
\text{Since it is positive, the relation is also positive.  }
$$
#### 3




#### 5.3.2

1. (1 points) Let $c_1$, $c_2$, and $c_3$ be the confidence values of the rules $ \text{ {a } } \rightarrow   \text{{b }} $, $ \text{ {a } } \rightarrow   \text{{b,c }} $, and $ \text{ {a,c } } \rightarrow   \text{{b }} $ respectively. If we assume that $c_1$, $c_2$, and $c_3$ have different values, what are the possible inequality relationships (e.g. $c_1 \leq c_2 \leq c_3$) among $c_1$, $c_2$, and $c_3$? Which rule has the lowest confidence?
2. (0.5 points) Suppose the confidence of the rules  $ \text{ {a } } \rightarrow   \text{{b }} $ and $ \text{ {b } } \rightarrow   \text{{c }} $ are larger than the confidence threshold. Is it possible that $ \text{ {a } } \rightarrow   \text{{c }} $ has a confidence below that threshold? If no, explain why. If yes, give an example. 

1.

Possible inequality relationships:

$c_{1} \leq c_{2} \leq c_{3}$

$c_{1} \leq c_{3} \leq c_{2}$

$c_{2} \leq c_{1} \leq c_{3}$

$c_{2} \leq c_{1} \leq c_{3}$

$c_{3} \leq c_{1} \leq c_{2}$

$c_{3} \leq c_{2} \leq c_{1}$

Confidence:
$$ \text{a } \rightarrow \text{b}\  \frac{\{a,b\}}{\{a\}}$$

$$ \text{a } \rightarrow \text{b,c }\  \frac{\{a,b,c\}}{\{a\}}$$

$$ \text{a,c } \rightarrow \text{b}\  \frac{\{a,b,c\}}{\{a,c\}}$$
,
Support $ \{a\} > \{a,b\}, \{a,c\} > \{a,b,c\}$

$\rightarrow \frac{\{a,b\}}{\{b\}} >\ \frac{\{a,b,c\}}{\{a\}} , \frac{\{a,b,c\}}{\{a.c\}}$

$\frac{\{a,b,c\}}{\{a,c\}} > \frac{\{a,b,c\}}{\{a\}}$


The rule with the lowest confidence is $c_{2}: \{a\} \rightarrow \{b,c\}$

2.

Lets say we have the following data set:

1.a,b

2.b,c

3.a,b

4.a,b

5.a,b

6.b,c

7.b,c

8.a,c


The confidence rule for $a \rightarrow b = \frac{4}{5} = 80\%$

The confidence rule for $b \rightarrow c = \frac{3}{3} = 100\%$

Then the confidence rule for $a \rightarrow \text{c becomes very small } \frac{1}{5}$ So only 20%

This dataset shows that even if the first two rules are above the threshold, a-> c does not necessarily have to be as well. Now this data set is very small and not seem very realistic, so the chances of it happening in a real life example are smaller, especially if there are many more item instances.





#### 5.3.3

(3 points) Consider the relationships between customers who buy high-definition televisions and exercise machines as shown in Table 2 and 3.

1. Compute the odd ratios for both tables.
2. Compute the $\phi$-coefficient for both tables.
3. Compute the interest (or lift, in the book) factor for both tables.

For Table 3 you should compute measures given above separately for College
Students and for Adults. For each of the measures, describe how the direction
of association changes when data is pooled together (Table 2) instead of being
separated into two groups (Table 3)

##### Table 2: Two way contingency table between the sale of high-definition television and exercise machine
| |   Buy Exercise machine |     |     |
| :------------- | -------------:| :-----------:| :----------:| 
| **Buy HDTV     ** | yes | no | total |
| yes  | 105| 87 | 192 | 
| no | 40| 62 | 102 |   
| total | 145 | 149 | 294 | 
 

##### Table 3: Example of three-way contingency table
| | |   Buy Exercise machine |     |     |
|--- | :------------- | -------------:| :-----------:| :----------:| 
|**Customer group** | **Buy HDTV     ** | yes | no | total |
|College students | yes  | 2| 9 | 11 | 
| | no | 5| 20 | 25 |
| Working adults | yes  | 103| 78 | 181 | 
| | no | 35| 42 | 77 |  



Odds ratio = 

Odds ratio table 2:
$$\frac{105\cdot62}{40\cdot87}= 1.87$$
Odds ratio table 3, adults:
$$\frac{2\cdot20}{5\cdot9}=0.889$$
Odds ratio table 3, students:
$$\frac{103\cdot42}{35\cdot78}=1.58$$

When pooled together, the data suggests that peaople with a TV also buy an exercise machine. While this is the case for the adults, the opposite is true for students: people without a TV buy an exercise machine.

Phi-coeffcient = $\frac{P\left(A,B\right)-P\left(A\right)\cdot P\left(B\right)}{\sqrt{P\left(A\right)\cdot P\left(B\right)\cdot\left(1-P\left(A\right)\right)\cdot\left(1-P\left(B\right)\right)}}$
Phi-coefficient table 2:
$$\frac{\left(\frac{105}{294}-\frac{40}{294}\cdot\frac{87}{294}\right)}{\sqrt{\frac{40}{294}\cdot\frac{87}{294}\cdot\left(1-\frac{40}{294}\right)\cdot\left(1-\frac{87}{294}\right)}}=2.02$$

Phi-coefficient table 3, adults:
$$\frac{\left(\frac{103}{258}-\frac{78}{258}\cdot\frac{35}{258}\right)}{\sqrt{\frac{78}{258}\cdot\frac{35}{258}\cdot\left(1-\frac{78}{258}\right)\cdot\left(1-\frac{35}{258}\right)}}=2.28$$
Phi-coefficient table 3, students:
$$\frac{\left(\frac{2}{36}-\frac{9}{36}\cdot\frac{5}{36}\right)}{\sqrt{\frac{9}{36}\cdot\frac{5}{36}\cdot\left(1-\frac{9}{36}\right)\cdot\left(1-\frac{5}{36}\right)}}= 0.139$$

Interest = $\frac{P\left(A,B\right)}{P\left(A\right)\cdot P\left(B\right)}$

Interest table 2:
$$\frac{\frac{105}{294}}{\frac{40}{294}\cdot\frac{87}{294}}=8.87$$
Interest table 3, adults:
$$\frac{\frac{103}{258}}{\frac{78}{258}\cdot\frac{35}{258}}=9.73$$
Interest table 3, students:
$$\frac{\frac{2}{36}}{\frac{5}{36}\cdot\frac{9}{36}}=1.6$$